servant-server/src/Servant/Server/Internal/Authentication.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE FlexibleInstances   #-}
+
+module Servant.Server.Internal.Authentication
+( AuthProtected (..)
+, AuthData (..)
+, AuthHandlers (AuthHandlers, onMissingAuthData, onUnauthenticated)
+, basicAuthLax
+, basicAuthStrict
+, laxProtect
+, strictProtect
+, jwtAuthHandlers
+, jwtAuthStrict
+        ) where
+
+import           Control.Monad              (guard)
+import qualified Data.ByteString            as B
+import           Data.ByteString.Base64     (decodeLenient)
+#if !MIN_VERSION_base(4,8,0)
+import           Data.Monoid                ((<>), mempty)
+#else
+import           Data.Monoid                ((<>))
+#endif
+import           Data.Proxy                 (Proxy (Proxy))
+import           Data.String                (fromString)
+import           Data.Word8                 (isSpace, toLower, _colon)
+import           GHC.TypeLits               (KnownSymbol, symbolVal)
+import           Data.Text.Encoding         (decodeUtf8)
+import           Data.Text                  (splitOn)
+import           Network.Wai                (Request, requestHeaders)
+import Servant.Server.Internal.ServantErr (err401, ServantErr(errHeaders))
+import           Servant.API.Authentication (AuthPolicy (Strict, Lax),
+                                             AuthProtected,
+                                             BasicAuth (BasicAuth),
+                                             JWTAuth(..))
+import            Web.JWT                    (decodeAndVerifySignature, JWT, VerifiedJWT, Secret)
+import qualified  Web.JWT as JWT             (decode)
+
+-- | Class to represent the ability to extract authentication-related
+-- data from a 'Request' object.
+class AuthData a where
+    authData :: Request -> Maybe a
+
+-- | handlers to deal with authentication failures.
+data AuthHandlers authData = AuthHandlers
+    {   -- we couldn't find the right type of auth data (or any, for that matter)
+        onMissingAuthData :: IO ServantErr
+    ,
+        -- we found the right type of auth data in the request but the check failed
+        onUnauthenticated :: authData -> IO ServantErr
+    }
+
+-- | concrete type to provide when in 'Strict' mode.
+data instance AuthProtected authData usr subserver 'Strict =
+    AuthProtectedStrict { checkAuthStrict :: authData -> IO (Maybe usr)
+                        , authHandlers :: AuthHandlers authData
+                        , subServerStrict :: subserver
+                        }
+
+-- | concrete type to provide when in 'Lax' mode.
+data instance AuthProtected authData usr subserver 'Lax =
+    AuthProtectedLax { checkAuthLax :: authData -> IO (Maybe usr)
```

Comment:
No need. You can define `checkAuthXXX` only when your connection pool is in scope.